File: python_code/p017_medium_letter_combination.py
class Solution:
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        if digits == '' or digits is None: return []
        from itertools import product
        # print(list(product('AB', '12')))  # [('A', '1'), ('A', '2'), ('B', '1'), ('B', '2')]
        num_letter_dict = {
            '2': 'abc', '3': 'def', '4': 'ghi', '5': 'jkl',
            '6': 'mno', '7': 'pqrs', '8': 'tuv', '9': 'wxyz'
        }
        letter_list = [num_letter_dict[digit] for digit in digits]
        result = product(*letter_list)
        return [''.join(tup) for tup in result]


if __name__ == '__main__':
    x = Solution()
    print(x.letterCombinations('23'))

File: java_code/p017_medium_letter_combination.java
import java.util.ArrayList;
import java.util.List;
import java.util.HashMap;

// Solution
class letterCombination {
    public List<String> letterCombinations(String digits) {
        List<String> result = new ArrayList<>();
        if (digits == null || digits.equals("")) return result;
        HashMap<Character, String> num2Letters = new HashMap<>();
        num2Letters.put('2', "abc");
        num2Letters.put('3', "def");
        num2Letters.put('4', "ghi");
        num2Letters.put('5', "jkl");
        num2Letters.put('6', "mno");
        num2Letters.put('7', "pqrs");
        num2Letters.put('8', "tuv");
        num2Letters.put('9', "wxyz");

        // init list
        String letters = num2Letters.get(digits.charAt(0));
        for (int i = 0; i < letters.length(); i++) {
            result.add(String.valueOf(letters.charAt(i)));
        }
        // loop combinations
        for (int i = 1; i < digits.length(); i++) {
            int size = result.size();
            List<String> tmp = new ArrayList<>();
            String nextLetters = num2Letters.get((digits.charAt(i)));
            for (int j = 0; j < size; j++) {
                String x = result.get(j);
                for (int k = 0; k < nextLetters.length(); k++) {
                    tmp.add(x + nextLetters.charAt(k));
                }
            }
            result = tmp;
        }
        return result;
    }

    public static void main(String[] args) {

    }
}
